kyo-data/shared/src/main/scala/kyo/internal/ConcreteTagMacro.scala

```diff
@@ -0,0 +1,102 @@
+package kyo.internal
+
+import kyo.Chunk
+import kyo.ConcreteTag
+import kyo.ConcreteTag.*
+import scala.quoted.*
+
+private[kyo] object ConcreteTagMacro:
+
+    def derive[A: Type](using Quotes): Expr[ConcreteTag[A]] =
+        import quotes.reflect.*
+
+        def checkType(tpe: TypeRepr) =
+            if !tpe.dealias.typeSymbol.isClassDef then
+                report.errorAndAbort(
+                    s"""This method requires a ConcreteTag, but the type ${tpe.show} is not a class.
+                       |Consider using a concrete class instead of an abstract type or type parameter.
+                       |For generic types, you can provide an evidence of ConcreteTag explicitly.
+                       |Example: def method[A](a: A)(using ConcreteTag[A]) = ???""".stripMargin
+                )
+            end if
+            if tpe.typeArgs.nonEmpty then
+                report.errorAndAbort(
+                    s"""This method requires a ConcreteTag, but the type ${tpe.show} has type parameters.
+                       |This is a current limitation that may be lifted in future versions.
+                       |For now, use a non-generic type.""".stripMargin
+                )
+            end if
+            if tpe =:= TypeRepr.of[Null] then
+                report.errorAndAbort(
+                    s"""This method requires a ConcreteTag, but Null is not a valid type for ConcreteTag.
+                       |ConcreteTag does not support Null to prevent runtime errors.""".stripMargin
+                )
+            end if
+        end checkType
+
+        def create(tpe: TypeRepr): Expr[ConcreteTag[Any]] =
+            tpe match
+                case OrType(_, _) =>
+                    def flatten(tpe: TypeRepr): Chunk[TypeRepr] =
+                        tpe match
+                            case OrType(a, b) => flatten(a).concat(flatten(b))
+                            case _            => Chunk(tpe)
+                    val exprs = flatten(tpe).map(create)
+                    '{ Union(Set(${ Varargs(exprs) }*)) }
+                case AndType(_, _) =>
+                    def flatten(tpe: TypeRepr): Chunk[TypeRepr] =
+                        tpe match
+                            case AndType(a, b) => flatten(a).concat(flatten(b))
+                            case _             => Chunk(tpe)
+                    val exprs = flatten(tpe).map(create)
+                    '{ Intersection(Set(${ Varargs(exprs) }*)) }
+                case _ => createSingle(tpe).asInstanceOf[Expr[ConcreteTag[Any]]]
```

Comment:
not sure if pre-existing but the `asInstanceOf` doesn't seem necessary here

Reply:
ah good catch